File: ta_lib/strategies/signal/src/zerocross/mad.rs
use base::prelude::*;
use core::prelude::*;
use timeseries::prelude::*;

pub struct MadZeroCrossSignal {
    source: SourceType,
    smooth: Smooth,
    period_fast: usize,
    period_slow: usize,
}

impl MadZeroCrossSignal {
    pub fn new(source: SourceType, smooth: Smooth, period_fast: f32, period_slow: f32) -> Self {
        Self {
            source,
            smooth,
            period_fast: period_fast as usize,
            period_slow: period_slow as usize,
        }
    }
}

impl Signal for MadZeroCrossSignal {
    fn lookback(&self) -> usize {
        std::cmp::max(self.period_fast, self.period_slow)
    }

    fn trigger(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let mad =
            data.source(self.source)
                .spread_pct(self.smooth, self.period_fast, self.period_slow);

        (mad.cross_over(&ZERO), mad.cross_under(&ZERO))
    }
}

File: factor/signal/zerocross/mad.py
from dataclasses import dataclass

from core.models.parameter import (
    Parameter,
    StaticParameter,
)
from core.models.smooth import Smooth
from core.models.source import SourceType
from factor.signal.base import Signal, SignalType


@dataclass(frozen=True)
class MadZeroCrossSignal(Signal):
    type: SignalType = SignalType.MadZeroCross
    source: Parameter = StaticParameter(SourceType.CLOSE)
    smooth: Parameter = StaticParameter(Smooth.SMA)
    period_fast: Parameter = StaticParameter(8.0)
    period_slow: Parameter = StaticParameter(23.0)
